File: 170. Two Sum III - Data structure design/TwoSum.py
class TwoSum:
  def __init__(self) -> None:
    self.numberCounts = {}

  def add(self, num):
    if num not in self.numberCounts:
      self.numberCounts[num] = 0
    
    self.numberCounts[num] += 1

  def find(self, target):
    for num in self.numberCounts:
      remainingSum = target - num
      if num == remainingSum:
        if self.numberCounts[remainingSum] > 1:
          return  True
      elif remainingSum in self.numberCounts:
        return True

    return False

solution = TwoSum()
solution.add(1)
solution.add(3)
solution.add(5)
solution.add(3)
solution.add(2)
solution.add(2)
print('numberCounts: ', solution.numberCounts)
print('find(4): ', solution.find(4))
print('find(7): ', solution.find(7))

File: 170. Two Sum III - Data structure design/TwoSum.js
class TwoSum {
  constructor() {
    this.numberCounts = {};
  }

  add(num) {
    if (!this.numberCounts[num]) {
      this.numberCounts[num] = 0;
    }

    this.numberCounts[num]++;
  }

  find(target) {
    for(let num in  this.numberCounts) {
      let remainingSum = target - num;

      if (num === remainingSum) {
        if (this.numberCounts[num] > 1) {
          return true;
        }
      }
      
      if (this.numberCounts[remainingSum]) {
        return true;
      }
    }

    return false;
  }
}

const twoSum = new TwoSum();
twoSum.add(1);
twoSum.add(3);
twoSum.add(5);
twoSum.add(3);
twoSum.add(2);
twoSum.add(2);
console.log('numberCounts: ', twoSum.numberCounts);
console.log('find(4): ', twoSum.find(4));
console.log('find(7): ', twoSum.find(7));